### PR TITLE
[FrameworkBundle] Fall back to default configuration in debug:config and consistently resolve parameter values

### LINKED ISSUE #41121: `bin/console debug:config XXXX` does't work when the config is not defined in the **application**
**Symfony version(s) affected**: 4.4.*

**Description**

```
root /app bin/console debug:config JMSSerializerBundle

In ConfigDebugCommand.php line 92:
                                                                          
  The extension with alias "jms_serializer" does not have configuration.  
                                                                          

debug:config [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command> [<name> [<path>]]
```

**How to reproduce**

Drop all the configuration of a bundle

**Possible Solution**

Dump the real configuration, even if the app definition does not exists.

**Additional context**

```
root /app bin/console debug:config --help
Description:
  Dumps the current configuration for an extension
```

The help says it dump the **current**configuration, it means the configuration defined in the application merged with the default one (I know it, I made this feature).

This line cause the issue: https://github.com/symfony/symfony/blob/18e6e2d1599a2a7b7e24dcec64f0aeee67f235aa/src/Symfony/Component/DependencyInjection/Compiler/ValidateEnvPlaceholdersPass.php#L73

I can submit a PR to remove it if you are OK

### PR DESCRIPTION
**Pull request by herndlm**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #41121
| License       | MIT
| Doc PR        | n/a

Brings back determining the default configuration from pull request #30648. Refactors determining the extension config into a helper method to keep cognitive load low with early exits instead of nested ifs.

Additionaly, while reviewing, we noticed that entries with parameters are not resolved correctly if they come from the default config and an application config is existing but they are not overriden. E.g. the session config for the framework would output `save_path: %kernel.cache_dir%/sessions` instead of the resolved value.

Tested via unit test (obviously) and by manually removing bundle configuration in a Symfony project.

Beware that I do not know much about bundles and internals, I looked that up while implementing this and might have missed better/simpler solutions or edge cases. Please somebody also do another functional test to ensure this is really working as it should.

### CODE REVIEW

**Comment by herndlm** [comment]
The failing test should be unrelated.

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php [hunk 1]
```diff
@@ -153,4 +144,36 @@ private function getConfigForPath(array $config, string $path, string $alias)
 
         return $config;
     }
+
+    /**
+     * @return mixed
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php]: this should be array as per `getConfigForPath` (IMHO we can enforce a real type)

**Review by ro0NL — COMMENTED** [review]
looks generally OK to me

previously we resolved values (`resolveValue`) after processing => https://github.com/symfony/symfony/pull/30648/files#diff-6f37907eb60dac18aec41358f427dab3c34f1c3d6d71247506ee52a856a05a2dL91, but im not sure it's still applicable for just default config now.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/DefaultConfigTestBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+class Configuration implements ConfigurationInterface
+{
+    public function getConfigTreeBuilder(): TreeBuilder
+    {
+        $treeBuilder = new TreeBuilder('default_config_test');
+
+        $treeBuilder->getRootNode()
+            ->children()
+                ->scalarNode('foo')->defaultValue('bar')->end()
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/DefaultConfigTestBundle/DependencyInjection/Configuration.php]: `defaultValue('%kernel.debug%')` perhaps to see if it behaves the same as if it were configured

**Comment by herndlm** [comment]
> looks generally OK to me
> 
> previously we resolved values (`resolveValue`) after processing => https://github.com/symfony/symfony/pull/30648/files#diff-6f37907eb60dac18aec41358f427dab3c34f1c3d6d71247506ee52a856a05a2dL91, but im not sure it's still applicable for just default config now.

Good catch! This is about env vars, right? I'll try to add a test and adapt that.

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php (hunk 1 above)
Comment (herndlm) [reply src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php]: I did not trust those methods and their phpdocs enough :) will take another look tomorrow

Comment (ro0NL) [reply src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php]: it's a hard type ;) https://github.com/symfony/symfony/blob/cb5e72890b871c896d52abb526cd7ee375f11db1/src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php#L133

**Comment by ro0NL** [comment]
> This is about env vars, right

yes :D another node with default env would be nice i guess, but AFAIK that's already taken care of due the outer `resolveEnvPlaceholders` call already

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/DefaultConfigTestBundle/DependencyInjection/Configuration.php (hunk 2 above)
Comment (herndlm) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/DefaultConfigTestBundle/DependencyInjection/Configuration.php]: Added a different default value with a custom parameter. And this actually showed that the `resolveValue` you mentioned was missing and needed.

**Comment by herndlm** [comment]
> > This is about env vars, right
> 
> yes :D another node with default env would be nice i guess, but AFAIK that's already taken care of due the outer `resolveEnvPlaceholders` call already

I added another node to check that and yes you're right - it's already taken care of

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php [hunk 3]
```diff
@@ -153,4 +144,33 @@ private function getConfigForPath(array $config, string $path, string $alias)
...
+        $extensionConfig = [];
+        foreach ($container->getCompilerPassConfig()->getPasses() as $pass) {
+            if ($pass instanceof ValidateEnvPlaceholdersPass) {
+                $extensionConfig = $pass->getExtensionConfig();
+                break;
+            }
+        }
+
+        if (isset($extensionConfig[$extensionAlias])) {
+            return $extensionConfig[$extensionAlias];
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php:165]: now im unsure if `resolveValue()` was already missing here :) eg. config is given, but also inherits a default node which takes a parameter/env as default value

feel free to test :)

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php (hunk 3 above)
Comment (herndlm) [reply src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php:165]: you're right, this is inconsistent. I wonder if I should fix it here or create a follow-up PR.
Update: fixed it..

**Comment by herndlm** [comment]
@ro0NL sorry to bother you again, 3 questions, no worries if you can't answer all of them
1. After your approval I did another change making the value resolving consistent as you mentioned in https://github.com/symfony/symfony/pull/42368#discussion_r683728000, is that fine with you?
2. Do you think the failing redis test could be related or can I test this somehow? I was sure it's unrelated but it constantly fails
3. Is the core team regularly checking PRs like this (e.g. via label) or should somebody be tagged? I want to avoid pinging them unnecessarily

**Comment by ro0NL** [comment]
1) :ok_hand: 
2) im 99.99% sure failing redis connection is unrelated, however https://ci.appveyor.com/project/fabpot/symfony/builds/40325532 looks related (outdated fixture perhaps?)
3) cc @lyrixx :)

**Comment by herndlm** [comment]
Grr, so apparently one of the values to check has single ticks on Windows but not on Linux xD I started to ignore appveyor already because it randomly failed..
UPDATE: should be fixed
UPDATE2: installed redis extension and started a redis Docker instance and tests were running fine 🤷‍♂️
```sh
➜  symfony git:(bugfix/fallback-to-default-config) ✗ php ./phpunit src/Symfony/Bundle/FrameworkBundle
#!/usr/bin/env php
PHPUnit 9.5.7 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

Testing /Users/herndlm/Development/source/git-forks/symfony/src/Symfony/Bundle/FrameworkBundle
.............................................................   61 / 1427 (  4%)
.............................................................  122 / 1427 (  8%)
.............................................................  183 / 1427 ( 12%)
.............................................................  244 / 1427 ( 17%)
.............................................................  305 / 1427 ( 21%)
.............................................................  366 / 1427 ( 25%)
.............................................................  427 / 1427 ( 29%)
.............................................................  488 / 1427 ( 34%)
.............................................................  549 / 1427 ( 38%)
......................................................SS.....  610 / 1427 ( 42%)
.............................................................  671 / 1427 ( 47%)
.............................................................  732 / 1427 ( 51%)
.............................................................  793 / 1427 ( 55%)
.............................................................  854 / 1427 ( 59%)
.............................................................  915 / 1427 ( 64%)
.............................................................  976 / 1427 ( 68%)
............................................................. 1037 / 1427 ( 72%)
............................................................. 1098 / 1427 ( 76%)
............................................................. 1159 / 1427 ( 81%)
............................................................. 1220 / 1427 ( 85%)
............................................................. 1281 / 1427 ( 89%)
............................................................. 1342 / 1427 ( 94%)
............................................................. 1403 / 1427 ( 98%)
........................                                      1427 / 1427 (100%)

Time: 00:54.738, Memory: 240.50 MB

OK, but incomplete, skipped, or risky tests!
Tests: 1427, Assertions: 3427, Skipped: 2.

Legacy deprecation notices (156)
➜  symfony git:(bugfix/fallback-to-default-config) ✗ php -v
PHP 7.4.22 (cli) (built: Jul 29 2021 18:27:37) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Xdebug v3.0.4, Copyright (c) 2002-2021, by Derick Rethans
    with Zend OPcache v7.4.22, Copyright (c), by Zend Technologies
➜  symfony git:(bugfix/fallback-to-default-config) ✗ git rev-parse HEAD
0939cdc44eeff42d04b4817f876494585cbecd87
➜  symfony git:(bugfix/fallback-to-default-config) ✗
```
should be all good then

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/ConfigDebugCommandTest.php [hunk 4]
```diff
@@ -58,6 +59,16 @@ public function testParametersValuesAreResolved()
         $this->assertStringContainsString('secret: test', $tester->getDisplay());
     }
 
+    public function testDefaultParameterValueIsResolvedIfConfigIsExisting()
+    {
+        $tester = $this->createCommandTester();
+        $ret = $tester->execute(['name' => 'framework']);
+
+        $this->assertSame(0, $ret, 'Returns 0 in case of success');
+        $kernelCacheDir = self::$kernel->getContainer()->getParameter('kernel.cache_dir');
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/ConfigDebugCommandTest.php]: `$this->application->getKernel()` would work isnt it? Let's try to not introduce extra state :)

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/ConfigDebugCommandTest.php [hunk 5]
```diff
@@ -74,6 +85,33 @@ public function testDumpWithPrefixedEnv()
...
+        $tester = $this->createCommandTester();
+        $ret = $tester->execute(['name' => 'DefaultConfigTestBundle']);
+
+        $this->assertSame(0, $ret, 'Returns 0 in case of success');
+        $this->assertStringContainsString("baz: '%env(BAZ)%'", $tester->getDisplay());
+    }
+
+    public function testDumpThrowsExceptionWhenDefaultConfigFallbackIsImpossible()
+    {
+        $this->expectException(LogicException::class);
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/ConfigDebugCommandTest.php]: `\LogicException` (we dont `use` root symbols)

**Comment by herndlm** [comment]
@lyrixx what do you think?

**Comment by lyrixx** [comment]
If it works as expected, let's ship it

**Comment by herndlm** [comment]
Rebased on latest 4.4 to get rid of the appveyor fails. 7.2 and 7.4 failing tests should be unrelated, 7.2 tests were still working a week ago 🤔 

Shall this be merged then or are any further adaptions needed/wanted? @ro0NL @lyrixx

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @herndlm.
